lib/Transforms/LayoutPropagation/LayoutPropagation.cpp

```diff
@@ -0,0 +1,685 @@
+    for (auto operand : op->getOperands()) {
+      std::string mapStr;
+      llvm::raw_string_ostream os(mapStr);
+      AffineMap operandLayout;
+      if (assignedLayouts.contains(operand))
+        operandLayout = assignedLayouts.at(operand);
+      operandLayout.print(os);
+      note << "\n- Operand: " << operand << "; Layout: " << os.str();
+    }
+  });
+
+  TypeSwitch<Operation *>(op)
+      // Ops with special rules
+      .Case<ReduceOp>(
+          [&](auto op) { return rectifyIncompatibleOperandLayouts(op); })
+      .Default([&](Operation *op) {
+        // Default target layout is chosen arbitrarily as the first operand's
+        // layout for now. A different pass is responsible for optimizing the
+        // placement and mechanics of the layout conversion ops.
+        mlir::IRRewriter builder(&getContext());
+        const auto it = llvm::find_if(op->getOperands(), [this](Value pair) {
+          return assignedLayouts.contains(pair);
+        });
+        AffineMap targetLayout = assignedLayouts.at(*it);
+
+        for (auto &opOperand : op->getOpOperands()) {
+          if (!assignedLayouts.contains(opOperand.get())) continue;
+          AffineMap sourceLayout = assignedLayouts.at(opOperand.get());
+
+          if (sourceLayout != targetLayout) {
+            builder.setInsertionPoint(op);
+            ConvertLayoutOp convertOp = builder.create<ConvertLayoutOp>(
+                op->getLoc(), opOperand.get(), AffineMapAttr::get(sourceLayout),
+                AffineMapAttr::get(targetLayout));
+
+            // Layout of the result is the same as the target layout of the
+            // conversion. Mostly this is done for consistency: all ops have an
+            // attribute describing the layout of their results.
+            OpBuilder builder(&getContext());
+            assignedLayouts.insert({convertOp.getResult(), targetLayout});
+            setResultLayoutAttr(convertOp);
+            op->setOperand(opOperand.getOperandNumber(), convertOp.getResult());
+          }
+        }
+      });
+}
+
+void LayoutPropagation::rectifyIncompatibleOperandLayouts(ReduceOp op) {
+  mlir::IRRewriter builder(&getContext());
+  builder.setInsertionPoint(op);
+
+  for (const auto &[input, init] : llvm::zip(op.getInputs(), op.getInits())) {
+    AffineMap inputLayout = assignedLayouts.at(input);
+    AffineMap initLayout = assignedLayouts.at(init);
+    AffineMap reducedInputLayout =
+        convertLayoutForReduce(inputLayout, op.getDimensions());
+
+    if (reducedInputLayout != initLayout) {
+      ConvertLayoutOp convertOp = builder.create<ConvertLayoutOp>(
+          op->getLoc(), init, AffineMapAttr::get(initLayout),
+          AffineMapAttr::get(reducedInputLayout));
+      Value toReplace = convertOp.getResult();
+      // I'd like to just call op.setOperand(i, toReplace) here, but I can't
+      // figure out how the i'th entry in `getInits` corresponds to the general
+      // OpOperand index.
```

Comment:
it's a bit of a hack but you can use `op.getODSOperandIndexAndLength(1).first + i`

Reply:
I removed the comment; I think the replacement method is safer.